src/net/socket_address.rs

```diff
@@ -0,0 +1,81 @@
+#![allow(unsafe_code)]
+use core::mem::size_of;
+
+use crate::backend::{
+    c,
+    net::write_sockaddr::{encode_sockaddr_v4, encode_sockaddr_v6},
+};
+
+use super::{SocketAddr, SocketAddrV4, SocketAddrV6};
+
+#[cfg(unix)]
+use super::SocketAddrUnix;
+
+/// Opaque type equivalent to `sockaddr` in C.
+///
+/// This is always used behind a raw pointer that is cast from a pointer to a
+/// `sockaddr`-compatible C type, and then cast back to a `sockaddr` pointer
+/// to be passed to a system call.
+#[repr(C)]
+pub struct SockAddrRaw {
+    _data: [u8; 0],
+}
+
+/// A trait abstracting over the types that can be passed as a `sockaddr`.
+///
+/// Safety: implementers of this trait must ensure that `with_sockaddr` calls
+/// `f` with a pointer that is readable for the passed length, and points
+/// to data that is a valid socket address for the system calls that accept
+/// `sockaddr` as a const pointer.
+pub unsafe trait SockAddr {
```

Comment:
And, what would you think of renaming this to `SocketAddrArg`? `Sock` -> `Socket` for the same reason as above, and `Arg` since it is for describing arguments; return values will need to use some other mechanism as you noted in the top comment.